File: js/problem46.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

const swapValues = function(nums, i, j) {
    if (i === j) {
        return;
    }
    let temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
};

const permuteUtil = function(nums, start, res) {
    if (start === nums.length - 1) {
        res.push(nums.slice());
        return;
    }
    
    for (let i = start; i < nums.length; i++) {
        swapValues(nums, start, i);
        permuteUtil(nums, start + 1, res);
        swapValues(nums, start, i);
    }
};

var permute = function(nums) {
    if (nums.length === 0) {
        return [];
    } else if (nums.length === 1) {
        return [nums];
    }
    
    let res = [];
    
    permuteUtil(nums, 0, res);
    
    return res;
};

File: leetcodeJava/src/main/java/Problem46.java
// 46. Permutations

class Solution {
    
    private void addSolution(int[] nums, List<List<Integer>> res) {
        List<Integer> temp = new ArrayList<Integer>();
        for (int n : nums) {
            temp.add(n);
        }
        res.add(temp);
    }
    
    private void backtrack(int[] nums, int start, List<List<Integer>> res) {
        if (start == nums.length - 1) {
            this.addSolution(nums, res);
            return;
        }
        for (int i = start; i < nums.length; i++) {
            int temp = nums[start];
            nums[start] = nums[i];
            nums[i] = temp;
            this.backtrack(nums, start + 1, res);
            temp = nums[start];
            nums[start] = nums[i];
            nums[i] = temp;
        }
    }
    
    public List<List<Integer>> permute(int[] nums) {
        if (nums.length == 0) {
            return new ArrayList<List<Integer>>();
        }
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        this.backtrack(nums, 0, res);
        return res;
    }
}